File: 04_Microbot/Python/robot.py
#robot-johno.py
#perform robot actions from received commands

from microbit import *
import radio

class MotoBit:
    moto_l = 0x21
    moto_r = 0x20
    moto_on = 0x70
    
    def __init__(self, address = 0x59):
        self.ADDR = address
          
    def write16(self,a,b):
        i2c.write(self.ADDR, bytes([a,b]), repeat=False)
    
    # True or False
    def enable(self, pwr):
        if pwr:
            self.write16(0x70,1)
        else:
            self.write16(0x70,0)
            
    # 0 for right, 1 for left, speed -127 to 127    
    def set_speed(self, motor, speed):
        motor = motor + 32
        if speed>=0:
            self.write16(motor,128 + speed)
        else:
            speed = speed + 127
            self.write16(motor, speed)
    # left and right speeds
    def drive(self,left,right):
        self.set_speed(0,right)
        self.set_speed(1,left)



radio_group = 0 #number of radio frequency to use
group_set = False #has the radio group been set?
motobit = MotoBit()
motobit.enable(True) # Enable motor driver
radio.on()
radio.config(power=7)

stop_sign = Image(  "09990:"
                    "90099:"
                    "90909:"
                    "99009:"
                    "09990")

while True:# do forever
    if not group_set:
        display.show(str(radio_group))
        if button_a.is_pressed() and button_b.is_pressed():
            group_set = True
            radio.config(channel = radio_group)
            display.show(Image.HAPPY)
            
        elif button_a.is_pressed():
            if radio_group > 0:
                radio_group = radio_group - 1
                
        elif button_b.is_pressed():
            if radio_group < 255:
                radio_group = radio_group + 1 
        sleep(150)

    else:# radio group has been set
        incoming = radio.receive()
        
        if incoming is "forward":
            motobit.drive(127, 127)
            display.show(Image.ARROW_N)
        elif incoming is "stop":
            motobit.drive(0,0)
            display.show(stop_sign)
        elif incoming is "right":
            motobit.drive(127,0)
            display.show(Image.ARROW_E)
        elif incoming is "left":
            motobit.drive(0, 127)
            display.show(Image.ARROW_W)
        sleep(100)
        
        

File: 04_Microbot/Javascript/robot.js
let radioGroup = 0
let groupSet = 0
input.onButtonPressed(Button.AB, () => {
    if (groupSet != 1) {
        groupSet = 1
        radio.setGroup(radioGroup)
    }
})
radio.onDataPacketReceived( ({ receivedString: remoteString }) =>  {
    if (remoteString == "forward") {
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Forward, 100)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Forward, 100)
        basic.showLeds(`
            . . # . .
            . # # # .
            # . # . #
            . . # . .
            . . # . .
            `)
        basic.pause(200)
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Forward, 0)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Forward, 0)
    } else if (remoteString == "left") {
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Reverse, 50)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Forward, 50)
        basic.showLeds(`
            . . # . .
            . # . . .
            # # # # #
            . # . . .
            . . # . .
            `)
        basic.pause(200)
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Forward, 0)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Forward, 0)
    } else if (remoteString == "right") {
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Forward, 50)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Reverse, 50)
        basic.showLeds(`
            . . # . .
            . . . # .
            # # # # #
            . . . # .
            . . # . .
            `)
        basic.pause(200)
        motobit.setMotorSpeed(Motor.Left, MotorDirection.Forward, 0)
        motobit.setMotorSpeed(Motor.Right, MotorDirection.Forward, 0)
    } else {

    }
})
input.onButtonPressed(Button.A, () => {
    if (groupSet != 1 && radioGroup > 0) {
        radioGroup = radioGroup - 1
    }
})
input.onButtonPressed(Button.B, () => {
    if (groupSet != 1 && radioGroup < 255) {
        radioGroup = radioGroup + 1
    }
})
while (groupSet != 1) {
    basic.showNumber(radioGroup)
}
basic.showIcon(IconNames.Happy)
basic.pause(100)
basic.showNumber(radioGroup)
motobit.enable(MotorPower.On)
